FuelLabs/fuel-indexer
# feature: support for GraphQL list types

### Description

- PR adds support for `[List]` types in GraphQL schema
- Most of the changes in this PR are actually related to https://github.com/FuelLabs/fuel-indexer/issues/1031
  - I just converted that `enum` into our `T: Decoder` pattern
- Adds support for many-to-many relationships (or join tables)
  - This covers the first 50% of https://github.com/FuelLabs/fuel-indexer/issues/947
  - Still need to persist the data into the m2m table and allow for a way to query though
- Integrates last remaining types from #897 
  - The v1 block explorer is essentially finished!! 🥳 🥳 

### Some (really rough) stats
- Stats are just from a snapshot of my computer
- A 2022 M2 chip with 24GB RAM, 8 CPUs
```bash
> uname -a
Darwin [..] Darwin Kernel Version 22.1.0: [ ... ]/RELEASE_ARM64_T8112 arm64
```

- 32 blocks/second
- ~100% CPU
- 30+ threads
  - Why so many threads for a single executor? 
    - Are they sqlx threads?
- Data size (at Block#125060)
  - 874 MB / 125060 (blocks) = ~7KB per block
```bash
postgres=# SELECT schema_name,
       pg_size_pretty(sum(table_size)::bigint),
       (sum(table_size) / pg_database_size(current_database())) * 100
FROM (
  SELECT pg_catalog.pg_namespace.nspname as schema_name,
         pg_relation_size(pg_catalog.pg_class.oid) as table_size
  FROM   pg_catalog.pg_class
     JOIN pg_catalog.pg_namespace ON relnamespace = pg_catalog.pg_namespace.oid
) t
GROUP BY schema_name
ORDER BY schema_name;
    schema_name     | pg_size_pretty |          ?column?
--------------------+----------------+----------------------------
 fuel_explorer      | 874 MB         |    85.09221162849698165300
 information_schema | 88 kB          | 0.008364440305203270644300
 pg_catalog         | 7712 kB        |     0.73302913220145026400
 pg_toast           | 144 MB         |    13.97089652068179014100
 public             | 432 kB         |     0.04106179786190696500
(5 rows)
```
### Testing steps

- [ ] Tests were added so CI should pass
- [ ] Refresh DB `bash scripts/utils/refresh_test_db.bash`
- [ ] Build test indexers `bash scripts/utils/build_test_wasm_module.bash`
- [ ] Start local fuel node `cargo run -p fuel-node --bin fuel-node`
- [ ] Start local web-api. `cargo run -p web-api --bin web-api`
- [ ] Start service with test indexer 

```bash
cargo run --bin fuel-indexer -- run --manifest packages/fuel-indexer-tests/components/indices/fuel-indexer-test/fuel_indexer_test.yaml --run-migrations --verbose
```
- [ ] Try local GraphQL playground

```graphql
query {
  listtypeentity {
    required_all
    optional_inner
    optional_outer
    optional_all
    virtual_optional_inner
  }
}
```
Result should be:

```json
{
  "data": [
    {
      "optional_all": [
        1,
        3
      ],
      "optional_inner": [
        "hello",
        "world"
      ],
      "optional_outer": [
        1,
        2,
        3,
        4,
        5
      ],
      "required_all": [
        1,
        2,
        3
      ],
      "virtual_optional_inner": [
        {
          "name": "foo",
          "size": 1
        },
        {
          "name": "bar",
          "size": 2
        }
      ]
    }
  ]
}
```
- [ ] Test the explorer!

```bash
cargo run --bin fuel-indexer -- run --manifest examples/fuel-explorer/fuel-explorer/fuel_explorer.manifest.yaml --run-migrations --fuel-node-host beta-3.fuel.network --fuel-node-port 80
```
- [ ] Examples should work (dockerized and non-dockerized)

### Changelog

- feature: support for GraphQL list types
